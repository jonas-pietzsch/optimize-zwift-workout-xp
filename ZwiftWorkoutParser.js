import { XMLParser, XMLBuilder } from "fast-xml-parser";
import { ZwiftWorkout } from "./ZwiftWorkout.js";

const options = {
  ignoreAttributes: false,
  attributeNamePrefix: "attribute",
  preserveOrder: true,
};

export class ZwiftWorkoutParser {
  static parseZwiftWorkoutFile(fileBuffer) {
    const parser = new XMLParser(options);
    const parsedWorkoutFile = parser.parse(fileBuffer.toString());
    return new ZwiftWorkout(parsedWorkoutFile);
  }

  static assembleZwiftWorkoutFile(zwiftWorkout) {
    const builder = new XMLBuilder({
      ...options,
      format: true,
    });
    return builder.build(zwiftWorkout.contents);
  }
}
