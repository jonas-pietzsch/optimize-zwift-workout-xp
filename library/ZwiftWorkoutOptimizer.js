import cloneDeep from "lodash/cloneDeep.js";
import { buildOptions } from "./options.js";

export class ZwiftWorkoutOptimizer {
  static optimize(originalWorkout, customOptions) {
    const options = buildOptions(customOptions);
    const workout = cloneDeep(originalWorkout);
    const statistics = {
      steadyStateToIntervals: {
        amount: 0,
        durationSeconds: 0,
      },
      warmupToIntervals: {
        amount: 0,
        durationSeconds: 0,
      },
      cooldownToIntervals: {
        amount: 0,
        durationSeconds: 0,
      },
    };
    [statistics, workout, options].forEach(Object.freeze);

    const resultWorkoutBlocks = workout.trainingBlocks
      .map((block) => {
        if (block.SteadyState && options.steadyStateBlocks.optimize) {
          return ZwiftWorkoutOptimizer.transformSteadyStateBlockToIntervalBlock(
            block,
            statistics,
            options
          );
        } else if (
          (block.Warmup && options.warmupAndCooldownBlocks.optimizeWarmup) ||
          (block.Cooldown && options.warmupAndCooldownBlocks.optimizeCooldown)
        ) {
          return ZwiftWorkoutOptimizer.transformWarmupOrCooldownBlockToIntervalBlock(
            block,
            statistics,
            options
          );
        } else {
          return block;
        }
      })
      .flatMap((item) => item);

    workout.trainingBlocks = resultWorkoutBlocks;
    workout.name = workout.name += " (optimized)";
    workout.addTag("XP optimized");

    const optimizationResults =
      ZwiftWorkoutOptimizer.calculateOptimizationResult(statistics);
    if (!options.skipReportOutput) {
      ZwiftWorkoutOptimizer.reportOptimizationResults(optimizationResults);
    }
    return {
      statistics,
      optimizationResults,
      optimizedWorkout: workout,
    };
  }

  static transformSteadyStateBlockToIntervalBlock(block, statistics, options) {
    const {
      steadyStateBlocks: {
        minimumDurationSeconds,
        replacementBlocksDurationSeconds,
      },
    } = options;
    const { attributePower, attributeDuration, ...otherBlockAttributes } =
      block[":@"];
    const attributePowerParsed = Number.parseFloat(attributePower);
    const attributeDurationParsed = Number.parseInt(attributeDuration);

    if (attributeDurationParsed >= minimumDurationSeconds) {
      const amountOfIntervalBlockReplacements =
        attributeDurationParsed / replacementBlocksDurationSeconds;
      const amountOfWholeIntervalBlockReplacements = Math.floor(
        amountOfIntervalBlockReplacements
      );
      const replacementBlocks = Array.from(
        Array(amountOfWholeIntervalBlockReplacements).keys()
      ).map(() => ({
        IntervalsT: [],
        ":@": {
          ...otherBlockAttributes,
          attributeRepeat: "1",
          attributeOnDuration: `${replacementBlocksDurationSeconds / 2}`,
          attributeOnPower: `${attributePowerParsed}`,
          attributeOffDuration: `${replacementBlocksDurationSeconds / 2}`,
          attributeOffPower: `${attributePowerParsed}`,
        },
      }));

      const partialIntervalBlockReplacementIsRequired =
        amountOfIntervalBlockReplacements >
        amountOfWholeIntervalBlockReplacements;
      if (partialIntervalBlockReplacementIsRequired) {
        const partialDuration =
          replacementBlocksDurationSeconds *
          (amountOfIntervalBlockReplacements -
            amountOfWholeIntervalBlockReplacements);
        replacementBlocks.push({
          IntervalsT: [],
          ":@": {
            ...otherBlockAttributes,
            attributeRepeat: "1",
            attributeOnDuration: `${(partialDuration / 2).toFixed(0)}`,
            attributeOnPower: `${attributePowerParsed}`,
            attributeOffDuration: `${(partialDuration / 2).toFixed(0)}`,
            attributeOffPower: `${attributePowerParsed}`,
          },
        });
      }

      statistics.steadyStateToIntervals.amount++;
      statistics.steadyStateToIntervals.durationSeconds +=
        attributeDurationParsed;
      return replacementBlocks;
    } else {
      return block;
    }
  }

  static transformWarmupOrCooldownBlockToIntervalBlock(
    block,
    statistics,
    options
  ) {
    const { replacementBlocksDurationSeconds } =
      options.warmupAndCooldownBlocks;
    const blockAttributes = block[":@"];
    const {
      attributePowerHigh,
      attributePowerLow,
      attributeDuration,
      ...otherBlockAttributes
    } = blockAttributes;
    const attributePowerHighParsed = Number.parseFloat(attributePowerHigh);
    const attributePowerLowParsed = Number.parseFloat(attributePowerLow);
    const attributeDurationParsed = Number.parseInt(attributeDuration);
    const warmupPowerDelta = attributePowerHighParsed - attributePowerLowParsed;

    if (
      attributeDurationParsed >=
      options.warmupAndCooldownBlocks.minimumDurationSeconds
    ) {
      const amountOfIntervalBlockReplacements =
        attributeDurationParsed / replacementBlocksDurationSeconds;
      const amountOfWholeIntervalBlockReplacements = Math.floor(
        amountOfIntervalBlockReplacements
      );
      const replacementBlocks = Array.from(
        Array(amountOfWholeIntervalBlockReplacements).keys()
      ).map((blockIndex) => {
        const onSegmentIndex = blockIndex * 2;
        const offSegmentIndex = blockIndex * 2 + 1;
        const onPower =
          attributePowerLowParsed +
          (onSegmentIndex / (amountOfWholeIntervalBlockReplacements * 2)) *
            warmupPowerDelta;
        const offPower =
          attributePowerLowParsed +
          (offSegmentIndex / (amountOfWholeIntervalBlockReplacements * 2)) *
            warmupPowerDelta;
        const duration = replacementBlocksDurationSeconds / 2;
        return {
          IntervalsT: [],
          ":@": {
            ...otherBlockAttributes,
            attributeRepeat: "1",
            attributeOnDuration: `${duration.toFixed(0)}`,
            attributeOnPower: `${onPower.toFixed(3)}`,
            attributeOffDuration: `${duration.toFixed(0)}`,
            attributeOffPower: `${offPower.toFixed(3)}`,
          },
        };
      });

      const partialIntervalBlockReplacementIsRequired =
        amountOfIntervalBlockReplacements >
        amountOfWholeIntervalBlockReplacements;
      if (partialIntervalBlockReplacementIsRequired) {
        const partialDuration =
          replacementBlocksDurationSeconds *
          (amountOfIntervalBlockReplacements -
            amountOfWholeIntervalBlockReplacements);
        replacementBlocks.push({
          IntervalsT: [],
          ":@": {
            ...otherBlockAttributes,
            attributeRepeat: "1",
            attributeOnDuration: `${(partialDuration / 2).toFixed(0)}`,
            attributeOnPower: `${attributePowerHighParsed}`,
            attributeOffDuration: `${(partialDuration / 2).toFixed(0)}`,
            attributeOffPower: `${attributePowerHighParsed}`,
          },
        });
      }

      if (block.Warmup) {
        statistics.warmupToIntervals.amount++;
        statistics.warmupToIntervals.durationSeconds += attributeDurationParsed;
      } else {
        statistics.cooldownToIntervals.amount++;
        statistics.cooldownToIntervals.durationSeconds +=
          attributeDurationParsed;
      }
      return replacementBlocks;
    } else {
      return block;
    }
  }

  static calculateOptimizationResult(statistics) {
    const result = {
      steadyStateToIntervals: {
        minutes: statistics.steadyStateToIntervals.durationSeconds / 60,
      },
      cooldownToIntervals: {
        minutes: statistics.cooldownToIntervals.durationSeconds / 60,
      },
      warmupToIntervals: {
        minutes: statistics.warmupToIntervals.durationSeconds / 60,
      },
    };
    result.steadyStateToIntervals.xp =
      result.steadyStateToIntervals.minutes * 2;
    result.cooldownToIntervals.xp = result.cooldownToIntervals.minutes * 6;
    result.warmupToIntervals.xp = result.warmupToIntervals.minutes * 6;
    result.totalXp = Object.values(result).reduce(
      (res, cur) => res + cur.xp,
      0
    );
    result.totalMinutes = Object.values(result).reduce(
      (res, cur) => res + (typeof cur === "object" ? cur.minutes : 0),
      0
    );
    return result;
  }

  static reportOptimizationResults(optimizationResults) {
    console.info(
      [
        `Workout optimization added ${optimizationResults.totalXp.toFixed(
          0
        )} XP!`,
        ` - ${optimizationResults.steadyStateToIntervals.minutes.toFixed(
          0
        )}m steady state blocks replaced with intervals: ${optimizationResults.steadyStateToIntervals.xp.toFixed(
          0
        )} extra XP`,
        ` - ${optimizationResults.cooldownToIntervals.minutes.toFixed(
          0
        )}m cooldown blocks replaced with intervals: ${optimizationResults.cooldownToIntervals.xp.toFixed()} extra XP`,
        ` - ${optimizationResults.warmupToIntervals.minutes.toFixed(
          0
        )}m warmup blocks replaced with intervals: ${optimizationResults.warmupToIntervals.xp.toFixed()} extra XP`,
      ].join("\n")
    );
  }
}
