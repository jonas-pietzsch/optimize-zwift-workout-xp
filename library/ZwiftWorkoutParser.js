import { XMLParser, XMLBuilder } from "fast-xml-parser";
import { ZwiftWorkout } from "./ZwiftWorkout.js";

const options = {
  ignoreAttributes: false,
  attributeNamePrefix: "attribute",
  preserveOrder: true,
};

export class ZwiftWorkoutParser {
  static xmlParser = new XMLParser(options);
  static xmlBuilder = new XMLBuilder({ ...options, format: true });

  static parseZwoFile(fileBuffer) {
    const parsedWorkoutFile = ZwiftWorkoutParser.xmlParser.parse(
      fileBuffer.toString()
    );
    return new ZwiftWorkout(parsedWorkoutFile);
  }

  static assembleZwoFile(zwiftWorkout) {
    return ZwiftWorkoutParser.xmlBuilder.build(zwiftWorkout.contents);
  }
}
