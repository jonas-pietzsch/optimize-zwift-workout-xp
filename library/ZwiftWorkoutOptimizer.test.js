import { ZwiftWorkoutParser } from "./ZwiftWorkoutParser.js";
import fs from "fs";
import cloneDeep from "lodash/cloneDeep.js";
import { ZwiftWorkoutOptimizer } from "./ZwiftWorkoutOptimizer.js";

describe("ZwiftWorkoutOptimizer", () => {
  describe("optimize", () => {
    let workout;

    beforeEach(() => {
      workout = ZwiftWorkoutParser.parseZwoFile(
        fs.readFileSync("./sample.zwo")
      );
    });

    it("should return the optimized workout", () => {
      const { optimizedWorkout } = ZwiftWorkoutOptimizer.optimize(workout, {
        skipReportOutput: true,
      });

      expect(optimizedWorkout.contents).toMatchSnapshot();
    });

    it("should not modify the original contents of the workout", () => {
      const originalSampleWorkout = cloneDeep(workout);

      const { optimizedWorkout } = ZwiftWorkoutOptimizer.optimize(workout, {
        skipReportOutput: true,
      });

      expect(workout.contents).toEqual(originalSampleWorkout.contents);
      expect(optimizedWorkout.contents).not.toEqual(workout.contents);
    });

    it("should return statistics and optimization results", () => {
      const { statistics, optimizationResults } =
        ZwiftWorkoutOptimizer.optimize(workout, {
          skipReportOutput: true,
        });

      expect(statistics).toEqual({
        cooldownOrWarmupToIntervals: {
          amount: 2,
          durationSeconds: 2000,
        },
        steadyStateToIntervals: {
          amount: 7,
          durationSeconds: 1620,
        },
      });
      expect(optimizationResults).toEqual({
        cooldownOrWarmupToIntervals: {
          minutes: 33.333333333333336,
          xp: 200,
        },
        steadyStateToIntervals: {
          minutes: 27,
          xp: 54,
        },
        totalXp: 254,
      });
    });
  });
});
