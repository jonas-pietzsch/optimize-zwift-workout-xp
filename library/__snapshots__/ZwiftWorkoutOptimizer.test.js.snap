// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ZwiftWorkoutOptimizer optimize should return the optimized workout 1`] = `
[
  {
    "workout_file": [
      {
        "author": [
          {
            "#text": "Somebody",
          },
        ],
      },
      {
        "name": [
          {
            "#text": "Something (optimized)",
          },
        ],
      },
      {
        "description": [
          {
            "#text": "Some description
Optimized on XP by https://jonas.verhoelen.de/optimize-zwift-workout-xp/",
          },
        ],
      },
      {
        "sportType": [
          {
            "#text": "bike",
          },
        ],
      },
      {
        "tags": [
          {
            ":@": {
              "attributename": "Some",
            },
            "tag": [],
          },
          {
            ":@": {
              "attributename": "Tags",
            },
            "tag": [],
          },
          {
            ":@": {
              "attributename": "XP optimized",
            },
            "tag": [],
          },
        ],
      },
      {
        "workout": [
          {
            ":@": {
              "attributeOffDuration": "60",
              "attributeOffPower": "0.464",
              "attributeOnDuration": "60",
              "attributeOnPower": "0.450",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "60",
              "attributeOffPower": "0.493",
              "attributeOnDuration": "60",
              "attributeOnPower": "0.479",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "60",
              "attributeOffPower": "0.521",
              "attributeOnDuration": "60",
              "attributeOnPower": "0.507",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "60",
              "attributeOffPower": "0.550",
              "attributeOnDuration": "60",
              "attributeOnPower": "0.536",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "60",
              "attributeOffPower": "0.579",
              "attributeOnDuration": "60",
              "attributeOnPower": "0.564",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "60",
              "attributeOffPower": "0.607",
              "attributeOnDuration": "60",
              "attributeOnPower": "0.593",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "60",
              "attributeOffPower": "0.636",
              "attributeOnDuration": "60",
              "attributeOnPower": "0.621",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "40",
              "attributeOffPower": "0.65",
              "attributeOnDuration": "40",
              "attributeOnPower": "0.65",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeDuration": "20",
              "attributePowerHigh": "1",
              "attributePowerLow": "0.7",
            },
            "Warmup": [],
          },
          {
            ":@": {
              "attributeCadence": "70",
              "attributeOffDuration": "60",
              "attributeOffPower": "0.6",
              "attributeOnDuration": "60",
              "attributeOnPower": "0.6",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeCadence": "70",
              "attributeOffDuration": "30",
              "attributeOffPower": "0.6",
              "attributeOnDuration": "30",
              "attributeOnPower": "0.6",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeCadence": "80",
              "attributeOffDuration": "60",
              "attributeOffPower": "0.7",
              "attributeOnDuration": "60",
              "attributeOnPower": "0.7",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeCadence": "80",
              "attributeOffDuration": "30",
              "attributeOffPower": "0.7",
              "attributeOnDuration": "30",
              "attributeOnPower": "0.7",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeCadence": "90",
              "attributeOffDuration": "60",
              "attributeOffPower": "0.8",
              "attributeOnDuration": "60",
              "attributeOnPower": "0.8",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeCadence": "90",
              "attributeOffDuration": "30",
              "attributeOffPower": "0.8",
              "attributeOnDuration": "30",
              "attributeOnPower": "0.8",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeCadence": "100",
              "attributeOffDuration": "60",
              "attributeOffPower": "0.9",
              "attributeOnDuration": "60",
              "attributeOnPower": "0.9",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeCadence": "100",
              "attributeOffDuration": "30",
              "attributeOffPower": "0.9",
              "attributeOnDuration": "30",
              "attributeOnPower": "0.9",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeCadence": "100",
              "attributeDuration": "20",
              "attributePower": "1.2",
            },
            "SteadyState": [],
          },
          {
            ":@": {
              "attributeCadence": "85",
              "attributeOffDuration": "60",
              "attributeOffPower": "0.55",
              "attributeOnDuration": "60",
              "attributeOnPower": "0.55",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeCadence": "85",
              "attributeOffDuration": "60",
              "attributeOffPower": "0.55",
              "attributeOnDuration": "60",
              "attributeOnPower": "0.55",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeCadence": "85",
              "attributeOffDuration": "30",
              "attributeOffPower": "0.55",
              "attributeOnDuration": "30",
              "attributeOnPower": "0.55",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "30",
              "attributeOffPower": "0.53",
              "attributeOnDuration": "30",
              "attributeOnPower": "1.33",
              "attributeRepeat": "10",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "60",
              "attributeOffPower": "0.55",
              "attributeOnDuration": "60",
              "attributeOnPower": "0.55",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "60",
              "attributeOffPower": "0.55",
              "attributeOnDuration": "60",
              "attributeOnPower": "0.55",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "30",
              "attributeOffPower": "0.55",
              "attributeOnDuration": "30",
              "attributeOnPower": "0.55",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "30",
              "attributeOffPower": "0.53",
              "attributeOnDuration": "30",
              "attributeOnPower": "1.33",
              "attributeRepeat": "10",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "60",
              "attributeOffPower": "0.55",
              "attributeOnDuration": "60",
              "attributeOnPower": "0.55",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "60",
              "attributeOffPower": "0.55",
              "attributeOnDuration": "60",
              "attributeOnPower": "0.55",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "30",
              "attributeOffPower": "0.55",
              "attributeOnDuration": "30",
              "attributeOnPower": "0.55",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "30",
              "attributeOffPower": "0.53",
              "attributeOnDuration": "30",
              "attributeOnPower": "1.33",
              "attributeRepeat": "10",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "60",
              "attributeOffPower": "0.619",
              "attributeOnDuration": "60",
              "attributeOnPower": "0.630",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "60",
              "attributeOffPower": "0.597",
              "attributeOnDuration": "60",
              "attributeOnPower": "0.608",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "60",
              "attributeOffPower": "0.574",
              "attributeOnDuration": "60",
              "attributeOnPower": "0.586",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "60",
              "attributeOffPower": "0.552",
              "attributeOnDuration": "60",
              "attributeOnPower": "0.563",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "60",
              "attributeOffPower": "0.530",
              "attributeOnDuration": "60",
              "attributeOnPower": "0.541",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "60",
              "attributeOffPower": "0.508",
              "attributeOnDuration": "60",
              "attributeOnPower": "0.519",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "60",
              "attributeOffPower": "0.486",
              "attributeOnDuration": "60",
              "attributeOnPower": "0.497",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "60",
              "attributeOffPower": "0.463",
              "attributeOnDuration": "60",
              "attributeOnPower": "0.474",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
          {
            ":@": {
              "attributeOffDuration": "60",
              "attributeOffPower": "0.441",
              "attributeOnDuration": "60",
              "attributeOnPower": "0.452",
              "attributeRepeat": "1",
            },
            "IntervalsT": [],
          },
        ],
      },
    ],
  },
]
`;
